src/test/java/com/proofpoint/discovery/TestIpHostnameAuthManager.java

```diff
@@ -0,0 +1,242 @@
+        assertThatIllegalArgumentException()
+                .isThrownBy(() -> new IpHostnameAuthManager(dynamicStore, configStore))
+                .withMessage("Unable to build discovery IP list");
+    }
+
+    @Test
+    public void testRead()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("127.0.0.1");
+        assertThatCode(() -> authManager.checkAuthRead(request)).doesNotThrowAnyException();
+    }
+
+    @Test
+    public void testReplicate()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("127.0.0.1");
+        assertThatCode(() -> authManager.checkAuthReplicate(request)).doesNotThrowAnyException();
+    }
+
+    @Test
+    public void testReplicateMismatch()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("10.20.30.40");
+        assertThatExceptionOfType(ForbiddenException.class)
+                .isThrownBy(() -> authManager.checkAuthReplicate(request))
+                .withMessageContaining("HTTP 403")
+                .withNoCause();
+    }
+
+    @Test
+    public void testReplicateInvalidIp()
+    {
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("10.10.10.10.10");
+        assertThatExceptionOfType(ForbiddenException.class)
+                .isThrownBy(() -> authManager.checkAuthReplicate(request))
+                .withMessageContaining("HTTP 403")
+                .withNoCause();
+    }
+
+    @Test
+    public void testDelete()
+    {
+        Id<Node> nodeId = Id.random();
+        DynamicAnnouncement toDelete = DynamicAnnouncement.copyOf(new DynamicAnnouncement("testing", "general", "/location", ImmutableSet.of(
+                new DynamicServiceAnnouncement(Id.random(), "storage", ImmutableMap.of("http", "http://localhost:4111"))
+        ))).setAnnouncer("10.20.30.40").build();
+        dynamicStore.put(nodeId, toDelete);
+
+        AuthManager authManager = new IpHostnameAuthManager(dynamicStore, configStore);
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getRemoteAddr()).thenReturn("10.20.30.40");
+        assertThatCode(() -> authManager.checkAuthDelete(nodeId, request)).doesNotThrowAnyException();
+        assertThatCode(() -> authManager.checkAuthDelete(Id.random(), request)).doesNotThrowAnyException();
```

Comment:
Test delete when no entry in `dynamicStore`?
Test delete when entry has null announcer?
Test delete when entry is a tombstone?

Reply:
The first two are resolved. I don't see a test of a delete when the entry is a tombstone.